From 77a39688df288fb808f96f32522aa6eded93c025 Mon Sep 17 00:00:00 2001
From: MengLi <meng.li@windriver.com>
Date: Sun, 7 Jul 2019 17:13:44 +0800
Subject: [PATCH 906/909] driver: ethernet: add shutdown fundtion for ravb
 driver on renesas rcar3 platform

When run kexec feature, ethernet device is not able to be opened
the second kernel boots up. Because the ethernet device is still in
running status. So, it is need to do correct release operation before
the second kernel boots up.
Therefore, add shutdown fundtion to stop the enternet device and
release corresponding resource.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/renesas/ravb_main.c | 26 ++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/drivers/net/ethernet/renesas/ravb_main.c b/drivers/net/ethernet/renesas/ravb_main.c
index ebe800777c7a..e988ce3b3950 100644
--- a/drivers/net/ethernet/renesas/ravb_main.c
+++ b/drivers/net/ethernet/renesas/ravb_main.c
@@ -2152,6 +2152,31 @@ static int ravb_remove(struct platform_device *pdev)
 	return 0;
 }
 
+static void ravb_shutdown(struct platform_device *pdev)
+{
+	struct net_device *ndev = platform_get_drvdata(pdev);
+	struct ravb_private *priv = netdev_priv(ndev);
+
+	/* Stop PTP Clock driver */
+	if (priv->chip_id != RCAR_GEN2)
+		ravb_ptp_stop(ndev);
+
+	dma_free_coherent(ndev->dev.parent, priv->desc_bat_size, priv->desc_bat,
+			  priv->desc_bat_dma);
+	/* Set reset mode */
+	ravb_write(ndev, CCC_OPC_RESET, CCC);
+	pm_runtime_put_sync(&pdev->dev);
+	unregister_netdev(ndev);
+	netif_napi_del(&priv->napi[RAVB_NC]);
+	netif_napi_del(&priv->napi[RAVB_BE]);
+	ravb_mdio_release(priv);
+	pm_runtime_disable(&pdev->dev);
+	free_netdev(ndev);
+	platform_set_drvdata(pdev, NULL);
+
+	return;
+}
+
 static int ravb_wol_setup(struct net_device *ndev)
 {
 	struct ravb_private *priv = netdev_priv(ndev);
@@ -2289,6 +2314,7 @@ static struct platform_driver ravb_driver = {
 		.pm	= &ravb_dev_pm_ops,
 		.of_match_table = ravb_match_table,
 	},
+	.shutdown	= ravb_shutdown,
 };
 
 module_platform_driver(ravb_driver);
-- 
2.17.1

